scripts/tuple_tree_generator/tuple_tree_generator/templates/struct_late.h.tpl

```diff
@@ -21,6 +21,16 @@ The notice below applies to the generated files.
 #include "revng/TupleTree/TupleLikeTraits.h"
 #include "revng/TupleTree/TupleTree.h"
 
+/** if root_type == struct.name -**/
+/** for dependency in struct.dependencies -**/
+/** if schema.get_definition_for(dependency).abstract -**/
+/** for child_type in schema.get_upcastable_types(schema.get_definition_for(dependency)) -**/
+#include "/*= user_include_path =*//*= child_type.name =*/.h"
+/** endfor **/
+/**- endif **/
+/**- endfor **/
+/**- endif **/
```

Comment:
It looks like invalidation added a whole lot of extra inter-dependencies. It kind of works now, since, as a "coincidence" everything that's needed already gets included by every translation unit anyway, but having something explicit like this makes me sleep a lot better at night.

Reply:
TL;DR for this is: "if this type depends of type X, include type X (delta some corner cases)"